src/pages/dataElements/edit/createJsonPatchOperations.ts

```diff
@@ -3,27 +3,42 @@ import { JsonPatchOperation } from '../../../types'
 import { DataElement } from '../../../types/generated'
 import { FormValues } from '../form/types'
 
-type DataElementKey = keyof DataElement
-
 interface FormatFormValuesArgs {
     dataElement: DataElement
     dirtyFields: { [name: string]: boolean }
     values: FormValues
 }
 
-const sanitizeDirtyValueKeys = (keys: DataElementKey[]) => {
-    const attributeValuesDirty = keys.find((key) =>
-        key.startsWith('attributeValues')
+const sanitizeDirtyValueKeys = (keys: string[]) => {
+    // these are removed from the dirtyKeys
+    // attributeValues is an array in the form, thus fields will be attributeValues[0] etc
+    // style.code should post to style, not style.code, because it's a complex object
+    const keyStartsWithToRemove = ['attributeValues', 'style'] as const
+    const shouldInclude = Object.fromEntries(
+        keys.map((key) => [key, false])
+    ) as Record<(typeof keyStartsWithToRemove)[number], boolean>
+
+    const keysWithout = keys.filter(
+        (key) =>
+            !keyStartsWithToRemove.some((keyToRemove) => {
+                const shouldRemove = key.startsWith(keyToRemove)
+                if (shouldRemove) {
+                    shouldInclude[keyToRemove] = true
+                }
+                return shouldRemove
+            })
     )
 
-    if (!attributeValuesDirty) {
+    // no difference
+    if (keysWithout.length === keys.length) {
         return keys
     }
 
-    return [
-        ...keys.filter((key) => !key.startsWith('attributeValues')),
-        'attributeValues' as DataElementKey,
-    ]
+    const keysToInclude = Object.entries(shouldInclude)
+        .filter(([, val]) => val)
+        .map(([key]) => key)
+
+    return keysWithout.concat(keysToInclude)
```

Comment:
This seems a little overly complex:

```ts
const complexKeys = ['attributeValues', 'style']
export const sanitizeDirtyValueKeys = (dirtyKeys: string[]) => {
    const complexChanges = complexKeys.filter(complexKey =>
        dirtyKeys.some(dirtyKey => dirtyKey.startsWith(complexKey))
    )
    const dirtyKeysWithoutComplexKeys = dirtyKeys.filter((dirtyKey) =>
        !complexKeys.some(complexKey => dirtyKey.startsWith(complexKey))
    )

    return [
        ...dirtyKeysWithoutComplexKeys,
        ...complexChanges,
    ]
}
```

I just checked, it seems to work just the same:

```ts
import { sanitizeDirtyValueKeys } from './createJsonPatchOperations'

describe('createJsonPatchOperations/sanitizeDirtyValueKeys', () => {
    it('should return the dirty values array as is', () => {
        const actual = sanitizeDirtyValueKeys(['foo', 'bar'])
        const expected = ['foo', 'bar']
        expect(actual).toEqual(expected)
    })

    it('should remove all attribute values changes and add a single "attributeValues"', () => {
        const actual = sanitizeDirtyValueKeys([
            'foo',
            'bar',
            'attributeValues[0].value',
            'attributeValues[1].value',
        ])
        const expected = ['foo', 'bar', 'attributeValues']
        expect(actual).toEqual(expected)
    })

    it('should remove style.icon and style.color changes and add a single "style"', () => {
        const actual = sanitizeDirtyValueKeys([
            'foo',
            'bar',
            'style.color',
            'style.icon',
        ])
        const expected = ['foo', 'bar', 'style']
        expect(actual).toEqual(expected)
    })
})
```

Reply:
Thanks! I will revisit this. Also made this in a hurry just before the demo. I think I was a bit too fixated of not having high runtime complexity. But `dirtKeys` is a fixed length, so shouldn't matter at all.